From 6b8ea698c3bc5e5db54c71d666dec594e81427ea Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 28 Apr 2025 13:25:32 +0800
Subject: [PATCH 095/115] mtk: mac80211: trigger a beacon update after setting
 beacon key

After setting the beacon key, a beacon update is needed if
IEEE80211_KEY_FLAG_GENERATE_MMIE is set on the key's flag so that the
beacon content contains the MMIE field.

Note that there is still time interval betweeen setting beacon key &
updating the beacon, so there might be some beacons with MMIE being
mis-written by HW

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 net/mac80211/key.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/net/mac80211/key.c b/net/mac80211/key.c
index b14e9cd9..70771d56 100644
--- a/net/mac80211/key.c
+++ b/net/mac80211/key.c
@@ -423,6 +423,10 @@ __ieee80211_set_default_beacon_key(struct ieee80211_link_data *link, int idx)
 
 	rcu_assign_pointer(link->default_beacon_key, key);
 
+	if (key && (key->conf.flags & IEEE80211_KEY_FLAG_GENERATE_MMIE))
+		ieee80211_link_info_change_notify(sdata, link,
+						  BSS_CHANGED_BEACON);
+
 	ieee80211_debugfs_key_update_default(sdata);
 }
 
-- 
2.45.2

